Source/JavaWrappers.cpp

```diff
@@ -274,6 +281,97 @@ namespace java::lang
+        : Object{android::global::GetWebSocketClass()}
+        , m_openCallback{std::move(open_callback)}
+        , m_messageCallback{std::move(message_callback)}
+        , m_closeCallback{std::move(close_callback)}
+        , m_errorCallback{std::move(error_callback)}
+    {
+        static JNINativeMethod methods[] = {
+                {"closeCallback", "()V", (void*)OnClose },
+                {"openCallback", "()V", (void*)&WebSocketClient::OnOpen },
+                {"messageCallback", "(Ljava/lang/String;)V", (void*)&WebSocketClient::OnMessage },
+                {"errorCallback", "()V", (void*)&WebSocketClient::OnError },
+        };
+        m_env->RegisterNatives(m_class, methods, 4);
+
+        // initialize socket
+        JObject(m_env->NewObject(m_class, m_env->GetMethodID(m_class, "<init>", "(Ljava/lang/String;)V"), m_env->NewStringUTF(url.c_str())));
+
+        WebSocketClient *p_webSocketClient = this;
+        s_objectMap.insert({JObject(), p_webSocketClient});
+
+        jmethodID connectSocket{m_env->GetMethodID(m_class, "connectBlocking", "()Z")};
+        m_env->CallBooleanMethod(JObject(), connectSocket);
+    }
+
+    void WebSocketClient::OnOpen(JNIEnv* env, jobject obj) 
+    {
+        auto itObject = WebSocketClient::FindClientInstance(env, obj);
+        if (itObject != WebSocketClient::s_objectMap.end()) 
+        {
+            itObject->second->m_openCallback();
+        }
+    }
+
+    void WebSocketClient::OnMessage(JNIEnv* env, jobject obj, jstring message) 
+    {
+        auto itObject = WebSocketClient::FindClientInstance(env, obj);
+        if (itObject != WebSocketClient::s_objectMap.end()) 
+        {
+            java::lang::String mystr{message};
+            itObject->second->m_messageCallback(mystr);
+        }
+    }
+
+    void WebSocketClient::OnClose(JNIEnv* env, jobject obj) 
+    {
+        auto itObject = WebSocketClient::FindClientInstance(env, obj);
+        if (itObject != WebSocketClient::s_objectMap.end()) 
+        {
+            itObject->second->m_closeCallback();
+
+            // delete the websocket object from map
+            WebSocketClient::s_objectMap.erase(itObject);
+        }
+    }
+
+    void WebSocketClient::OnError(JNIEnv* env, jobject obj) 
+    {
+        auto itObject = WebSocketClient::FindClientInstance(env, obj);
+        if (itObject != WebSocketClient::s_objectMap.end()) 
+        {
+            itObject->second->m_errorCallback();
+        }
+    }
+
+    std::unordered_map<jobject, java::websocket::WebSocketClient*>::iterator  WebSocketClient::FindClientInstance(JNIEnv* env, jobject obj) 
+    {
+        const auto it = std::find_if(WebSocketClient::s_objectMap.begin(), WebSocketClient::s_objectMap.end(), [&obj, &env](const auto &it)
+        {
+            return env->IsSameObject (obj, it.first);; // Comparing with the object
+        });
```

Comment:
Hmm, if we need to use find_if, then might as well not use a map at all :-). That said, you can give the std::unordered_map template argument a KeyEqual to help with this.

Reply:
Spoke offline with Richard, and have moved to std::vector.